### DISCUSSION TITLE
COW with ZFS or LVM

### DISCUSSION THREAD

**Discussion by ghost** [ltsp/ltsp#411] [Ideas]
We moved away from LTSP a few years ago as we were not to positive about the future of the project. We created a different set-up very specific to our use case so that we could continue to use our infrastructure the same way we did when we used LTSP. We based our system around ZFS or LVM volumes which get exported using iSCSI. We use [Lio](https://en.wikipedia.org/wiki/LIO_(SCSI_target)) for the exporting part.
I wrote the libraries for interaction with LVM, ZFS and Lio in Python. I wrote a small web interface in Node that helps manage the images and communicates with the Python daemon when a client wants to boot so that the required volume can be cloned and exported.
So, we us dnsmasq as a DHCP server or proxy and tftp server, a nodejs app for the configuration of the images which are either ZFS or LVM volumes and the Kernel to export the images. To maintain the images we use basic KVM/Libvirt. These images are COW which means that there is no need for any overlay.
I'm starting this discussion with the hope that there might be a way to merge my project into LTSP. I'm willing to contribute the libraries I've already written and I can possibly take on integration work with help from LTSP.

**Comment by alkisg** [comment]
Hi ghaoil, this is very interesting and a possible match with the direction that I have in mind for LTSP with https://github.com/ltsp/ltsp/issues/147. I'll start working on implementing a web service for LTSP next year though... Is your code available publicly somewhere?

> These images are COW which means that there is no need for any overlay.

How does that work, each client gets its own zfs/lvm pool or snapshot?

**Reply by ghost** [reply]
I will push the code online tomorrow and have a look at the issue you're referencing.

**Reply by ghost** [reply]
I started to upload some of the libraries. We usually distribute things in Debian packages though our private repository. The Debian packages are not confirmed to the Debian standard so I'm uploading their contents separately. That will allow me to document things at the same time.
